src/lower.c

```diff
@@ -1763,6 +1763,38 @@ static void lower_toggle_coverage(lower_unit_t *lu, tree_t decl)
    cover_pop_scope(lu->cover);
 }
 
+static void lower_state_coverage(lower_unit_t *lu, tree_t decl)
+{
+   assert(cover_enabled(lu->cover, COVER_MASK_STATE));
+
+   type_t type = tree_type(decl);
+   if (cover_skip_type_state(lu->cover, type))
+      return;
+
+   cover_push_scope(lu->cover, decl);
+
+   int hops = 0;
+   vcode_var_t var = lower_search_vcode_obj(decl, lu, &hops);
+   assert(var != VCODE_INVALID_VAR);
+
+   // Add single coverage tag per enum literal. This is to track literal string
+   // in the identifier of the coverage tag.
+   for (int i = 0; i < type_enum_literals(type); i++) {
+      tree_t literal = type_enum_literal(type, i);
```

Comment:
I don't think this works if `type` is a subtype of an enum. It should be like:

```c
type_t base = type_base_recur(type);
const int nlits = type_enum_literals(type);
for (int i = 0; i < nlits; i++) {
   tree_t literal = type_enum_literal(type, i);
```

Or perhaps just restrict the coverage items to the literals in the subtype:

```c
int64_t low, high;
if (!folded_bounds(range_of(type, 0), &low, &high))
   return;

type_t base = type_base_recur(type);
for (int i = low; i < high; i++) {
   tree_t literal = type_enum_literal(type, i);
```

Reply:
I see, using `type_base_recur` will help here for the subtype.

But the second counter-proposal won't work either. If I don't emit tags for literals starting from first (translated to 0x0), the run-time search for pointer to coverage data won't work correctly. To get it working, we would need to store index of the first literal in the tag, and then subtract it from actual enum value in the runtime code.

Reply:
> 
> But the second counter-proposal won't work either. If I don't emit tags for literals starting from first (translated to 0x0), the run-time search for pointer to coverage data won't work correctly. To get it working, we would need to store index of the first literal in the tag, and then subtract it from actual enum value in the runtime code.

OK let's just leave that for now.

Reply:
When you wrote it, I was half-way through implementation. I tried to give it a shot. 

I keep the `low` as an argument of the op, and I subtract from the index of the first tag upon callback setup.
This pre-compensates the tag index by a negative offset. Then, run-time CB adds the offset back, and gets
correct index of counter corresponding to first tag for given sub-type.

Reply:
Note that all this heavily relies on the fact that enums are numbered from zero in the generated code, and that sub-type enums have the same encoding as corresponding base type enums. I think it is reasonable to rely on that.

The current "addressing" of tag would completely fall apart if there would be a possibility to encode enum with non-monotonic values (e.g. with C/System Verilog like base-type). Luckily VHDL does not have this.